### PR TITLE
Implement the OAuth flow for the Setup CTA Banner.

### PR DESCRIPTION
**Pull request by ankitrox**
## Summary

Addresses issue:

- #8132 

## PR Author Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

**Review by techanvil — CHANGES_REQUESTED** [review]
Hi @ankitrox, this is generally looking good. I've left a few comments, please take a look. Please also note the comment regarding the QAB that I left on the linked issue: https://github.com/google/site-kit-wp/issues/8132#issuecomment-2150367693

File: assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationSetupCTAWidget.test.js [hunk 1]
```diff
@@ -383,5 +398,86 @@ describe( 'AudienceSegmentationSetupCTAWidget', () => {
...
+				);
+			} );
+
+			await waitFor( () =>
+				expect( enableAudienceGroupSpy ).toHaveBeenCalledTimes( 1 )
+			);
+			expect( queryByText( 'Enabling groups' ) ).toBeInTheDocument();
+		} );
+
+		it( 'should call the enableAudienceGroups action when `autosubmit` is set to true.', async () => {
```
Comment (techanvil) [inline assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationSetupCTAWidget.test.js:443]: Typo:

```suggestion
		it( 'should call the enableAudienceGroups action when `autoSubmit` is set to true.', async () => {
```

File: assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationSetupCTAWidget.test.js [hunk 2]
```diff
@@ -383,5 +398,86 @@ describe( 'AudienceSegmentationSetupCTAWidget', () => {
...
+				);
+			} );
+
+			await waitFor( () =>
+				expect( enableAudienceGroupSpy ).toHaveBeenCalledTimes( 1 )
+			);
+			expect( queryByText( 'Enabling groups' ) ).toBeInTheDocument();
+		} );
+
+		it( 'should call the enableAudienceGroups action when `autoSubmit` is set to true.', async () => {
```
Comment (techanvil) [inline assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationSetupCTAWidget.test.js:443]: Actually, the use of backticks is a little inconsistent here too. We might as well remove them from `autoSubmit`:

```suggestion
		it( 'should call the enableAudienceGroups action when autoSubmit is set to true.', async () => {
```

File: assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationSetupCTAWidget.test.js [hunk 3]
```diff
@@ -383,5 +393,46 @@ describe( 'AudienceSegmentationSetupCTAWidget', () => {
 
 			expect( queryByText( /Maybe later/i ) ).not.toBeInTheDocument();
 		} );
+
+		it( 'should call the enableAudienceGroups action when `Enable groups` CTA is clicked.', async () => {
```
[the remark this answers is not in the document]
Comment (techanvil) [reply assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationSetupCTAWidget.test.js:402]: As discussed on [Slack](https://10up.slack.com/archives/CBKKQEBR9/p1717602239368699), we'll stick with the current approach for now at least in order to unblock things.

**Review by techanvil — APPROVED** [review]
LGTM, nice one @ankitrox!

**Review by techanvil — CHANGES_REQUESTED** [review]
Hey @ankitrox, I had approved this but then spotted your [comment](https://github.com/google/site-kit-wp/issues/8132#issuecomment-2162133178) - I am not seeing the updated tests in the PR, do you need to push a change?

File: assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationSetupCTAWidget.test.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (techanvil) [reply assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationSetupCTAWidget.test.js:419]: Hey @ankitrox, thanks for updating the tests to take a more integrated approach.

However - the test descriptions still refer to the `enableAudienceGroups()` action. Please can you update them to remove the reference to the action (e.g. taking the suggestion in the comment above)?

File: assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationSetupCTAWidget.test.js [hunk 4]
```diff
@@ -383,5 +415,117 @@ describe( 'AudienceSegmentationSetupCTAWidget', () => {
...
+				status: 200,
+				body: settingsBody,
+			} );
+
+			muteFetch( reportEndpoint );
+
+			expect( queryByText( 'Enabling groups' ) ).toBeInTheDocument();
+		} );
+
+		it( 'should call the enableAudienceGroups action when autoSubmit is set to true.', async () => {
```
Comment (techanvil) [inline assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationSetupCTAWidget.test.js:475]: Please update the test description to remove the reference to `enableAudienceGroups`, for example:

```suggestion
		it( 'should initialise the list of configured audiences when autoSubmit is set to true.', async () => {
```

**Review by techanvil — CHANGES_REQUESTED** [review]
Thanks @ankitrox - just a last few tweaks and this looks good to go.

File: assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationSetupCTAWidget.test.js [hunk 5]
```diff
@@ -383,5 +415,119 @@ describe( 'AudienceSegmentationSetupCTAWidget', () => {
...
+			} );
+
+			muteFetch( reportEndpoint );
+
+			expect(
+				getByRole( 'button', { name: /Enabling groups/i } )
+			).toBeInTheDocument();
+		} );
+
+		it( 'should initialise the list of configured audiences when autoSubmit is set to true.', async () => {
```
Comment (techanvil) [inline assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationSetupCTAWidget.test.js:477]: I've noticed this test can fail sporadically - here it is [failing in CI](https://github.com/google/site-kit-wp/actions/runs/9482420631/job/26127263392?pr=8741), and I can reproduce the failure locally by running the test a few times:

image

We don't want to introduce a new flaky test; please can you take a look and see if you can fix it?

I had a quick look myself, hoping we could simply wrap the `waitForRegistry()` call in an `act()`, but that didn't fix it.

**Review by techanvil — CHANGES_REQUESTED** [review]
Hey @ankitrox, unfortunately I have spotted one more thing that needs addressing, please take a look.

File: assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationSetupCTAWidget.test.js (hunk 5 above)
Comment (techanvil) [reply assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationSetupCTAWidget.test.js:478]: Hey @ankitrox, further on this I've got what looks like a fix, so I've pushed that to save spending more time on this one.

**Review by techanvil — APPROVED** [review]
LGTM, nice one @ankitrox!

File: assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationSetupCTAWidget.test.js (hunk 5 above)
Comment (ankitrox) [reply assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationSetupCTAWidget.test.js:478]: Thank you so much @techanvil . It looks good now!

**Pull request merged**